routes/ecom/modules/list-payments.js

```diff
@@ -3,12 +3,49 @@
 // generate default payment gateway object
 const newPaymentGateway = require(process.cwd() + '/lib/new-payment-gateway')
 
+const checkDiscountItems = (paymentGateway, items) => {
+  const discount = paymentGateway.discount
+  const productIds = discount && discount.product_ids
+
+  if (!productIds || !productIds.length) return
+
+  if (items && items.length) {
+    let isWitoutDiscountItem = false
+    let discountValue = 0
+
+    items.forEach(item => {
+      if (productIds.includes(item.product_id)) {
+        discountValue += (item.quantity) * (item.final_price || item.price)
+      } else {
+        isWitoutDiscountItem = true
+      }
+    })
+
+    if (discountValue) {
+      if (discount.apply_at === 'subtotal' && discount.type === 'percentage' && discount.value > 0 && isWitoutDiscountItem) {
+        discountValue *= discount.value / 100
+        const newDiscount = {
+          apply_at: discount.apply_at,
+          value: discountValue,
+          type: 'fixed'
+        }
+        paymentGateway.discount = newDiscount
+      }
```

Comment:
```suggestion
      } else if (isWitoutDiscountItem) {
        delete paymentGateway.discount
      }
```

Reply:
https://github.com/ecomplus/app-paghiper/pull/15#discussion_r1675331755

Reply:
Sem isso aí você estaria mantendo o desconto, se não for percentual no subtotal, independente dos itens no carrinho baterem ou não com `productIds`

Reply:
Verdade, fez muito mais sentido do que da forma que usei.